Simulcast packet rewriting

daniel-abramov opened the pull request:
Fixes https://github.com/matrix-org/waterfall/issues/91
Fixes https://github.com/matrix-org/waterfall/issues/89
Fixes https://github.com/matrix-org/waterfall/issues/95

daniel-abramov committed:
simulcast: implement packet rewriting

daniel-abramov committed:
subscription: fix key frame requests

daniel-abramov committed:
tracker: delete subscriptions when peer is removed

daniel-abramov committed:
participant: remove canRequestKeyFrame logic

It seems like many key frame requests (i.e. when switching simulcast)
are quite critical and dropping them just because someone requested key
frame N milliseconds ago would result in us not delivering a key frame
that the participant may absolutely need to render the frame which would
result in participant seeing a grey frame for a while.

We probably can solve it in a more efficient way by e.g. storing the
key frames somewhere and firing them up (though they would be a bit too
old).

This is an experimental way to fix the issue.

daniel-abramov committed:
simulcast(rtp): properly handle edge cases

SimonBrandner commented on pkg/peer/subscription/rtp.go:
```diff
@@ -0,0 +1,151 @@
+	// Check if we've just switched the layer before this packet, i.e. if
+	// it is the first packet after switching layers.
+	if p.previouslyForwardedSSRC != p.selectedSSRC {
+		// There is a gap of 1 seqnum to signify to the decoder that the previous frame
+		// was (probably) incomplete. That's why there's a 2 for the seqnum.
+		delta := PacketIdentifiers{1, 2}
+
+		// We make an exception for the very first packet that we're forwarding
+		// so that we start with 0 seqnum and 0 timestamp for the first packet.
+		if p.previouslyForwardedSSRC == 0 {
+			delta = PacketIdentifiers{0, 0}
+		}
```
Can this be an if-else statement?

daniel-abramov replied:
I think I know what you mean :) I.e. if it was Rust I would have written it like this:

```rust
let delta = if p.previouslyForwardedSSRC == 0 {
    PacketIdentifiers::default()
} else {
    PacketIdentifiers::new(1, 2)
};
```

That's what you semantically would love to see, right?

If yes, then I'm also in favor of such initialization, but unfortunately in Go the only way to write it is either like this:

```go
var delta PacketIdentifiers
if p.previouslyForwardedSSRC == 0 {
    delta.sequenceNumber = 0, delta.timestamp = 0
} else {
    delta.sequenceNumber = 2, delta.timestamp = 1
}
```

Or, given that by default Go would initialize things with a default value (which is zero by default), shorter:

```go
var delta PacketIdenfiers
if p.previouslyForwadedSSRC != 0 {
    delta.sequenceNumber = 2, delta.timestamp = 1
}
```

The last one does not look that bad, but it seems like a variation of the current version? (except that the if is "flipped" and the initilization is default / less explicit).

Which one would you prefer?

SimonBrandner replied:
```go
var delta PacketIdentifiers
if p.previouslyForwardedSSRC == 0 {
    delta.sequenceNumber = 0, delta.timestamp = 0
} else {
    delta.sequenceNumber = 2, delta.timestamp = 1
}
```

^ reads the best from POV or perhaps (think that should work, right?):

```go
var delta PacketIdentifiers
if p.previouslyForwardedSSRC == 0 {
    delta = PacketIdentifiers{0, 0}
} else {
    delta = PacketIdentifiers{2, 1}
}
````

daniel-abramov replied:
Done.

daniel-abramov committed:
track: decouple simulcast info from track info

This allows us to treat them separatel and make simulcast something that
changes over time without really changing the track info. As a side
effect, we can treat them separately and lift the restriction in our
code that required that the track's initial SSRC (after publishing)
matches the SSRC of the packet.

It turned out that the SSRC of the published tracks can change over
time, i.e. after the renegotiation.

daniel-abramov committed:
conference: only use simulcast for video tracks

SimonBrandner commented on pkg/peer/peer.go:
```diff
@@ -99,16 +99,17 @@ func (p *Peer[ID]) SubscribeTo(track common.TrackInfo) (*subscription.Subscripti
+	}
 
-	rtcps := []rtcp.Packet{&rtcp.PictureLossIndication{MediaSSRC: ssrc}}
+	p.logger.Debugf("Keyframe request: %s (%s)", info.TrackID, info.Simulcast.Layer)
```
Yeah, perhaps

daniel-abramov replied:
This is not a bad idea, though I think if we move it to trace, we probably would never enable it? I.e. ideally key frame requests should not be sent that often, I would say that they are always logged in big quantities, it would signify a bug. During the testing with debug logs I've noticed that they don't really happen that often (primarily during the switch/subscribe/unsubscribe), otherwise, if they were logged too often, there were always some bugs that caused it.

Does it feel like they would not be that bad for the debug logs since we only would use them during the **debugging**? In production, I think we would anyway run the SFU with the default maximum level of `info`. What do you think?

I think tracing is more suitable for things like... if we let's say logged each packet (there would be hundreds of them per second), whereas keyframe requests do not even happen every second.

daniel-abramov replied:
Done.

dbkr commented on pkg/conference/participant/tracker.go:
```diff
@@ -200,19 +249,22 @@ func (t *Tracker) Unsubscribe(participantID ID, tracks []TrackID) {
 // Processes RTCP packets received on a given track.
-func (t *Tracker) ProcessKeyFrameRequest(info common.TrackInfo) error {
+func (t *Tracker) ProcessKeyFrameRequest(info common.TrackInfo, simulcast common.SimulcastLayer) error {
 	const sendKeyFrameInterval = 500 * time.Millisecond
```
This looks unused now?

daniel-abramov replied:
Oh, good catch, thanks! Golang linter unfortunately did not capture that.

dbkr commented on pkg/peer/subscription/rtp.go:
```diff
@@ -0,0 +1,135 @@
+			return nil, fmt.Errorf("Ignoring old packet")
+		}
+
+		delta := incomingIDs.Sub(p.firstIncomingIDs)
```
Bear in mind that these IDs will wrap around, SSRCs quite frequently, so this will end up being negative sooner or later.

daniel-abramov replied:
The timestamps and sequence numbers? Won't they be caught by the `if` above?

Sorry, I did not quite get what you mean by wrapping around SSRCs though.

dbkr replied:
Oh sorry, I meant seqnums not SSRCs.

daniel-abramov replied:
Do you by chance know what's the right way to rewrite these values once they wrap? I.e. what do actually browsers expect? I'm not that much concerned about the timestamps since they are `uint32` and so far I've never experienced them wrapping (though I believe it might happen for longer conferences?), but the wrapping of sequence numbers does happen for conferences e.g. conferences that last for half an hour.

Given that the sequence number is just an `uint16`, we reach this rather sooner (2^16=65536). Do you by chance know if the browser would expect a sequence number of `0` after we've sent the one with the sequence number `65535`?

dbkr replied:
Yeah, it just wraps around in the normal way, so 0 comes after 65535. Bear in mind that timestamps don't start from 0 though, so it's not just length of call that's the factor, it's also being unlucky and it choosing a starting TS near the top of the range.

dbkr commented on pkg/peer/subscription/video.go:
```diff
@@ -0,0 +1,145 @@
+		Timeout: 2 * time.Second,
+		OnTimeout: func() {
+			logger.Warnf("No RTP on subscription for %s (%s)", subscription.info.TrackID, subscription.currentLayer)
+			subscription.requestKeyFrame()
```
Was this going to switch down to whatever layer is still getting packets?

daniel-abramov replied:
Yeah, right, I've specifically created this helper `requestKeyFrame()` which uses `subscription.info` and `subscription.simulcast`, i.e. making sure that we use the actual current layer and not the layer that we used when creating the initial subscription.

dbkr replied:
ah, I missed that this was doing this

daniel-abramov committed:
minor: remove obsolete (unused) code

daniel-abramov committed:
subscription: make current layer atomic

It is accessed from several go-routines:
- from the "main" one which is called when we receive RTP frames or when
  we change the current layer
- from the RTCP reading routine
- from the watchdog routine

daniel-abramov committed:
subscription: don't use mutable pointer for RTP

This will modify the packet for everyone. We pass it by value, so that
the header is copied. The content wont' be copied as it contains arrays
that are only shallow copied.

daniel-abramov committed:
subscription: adjust logging (more practical)

Ignore requests come too often since oftentimes nothing changes, so
let's not log them as warnings, but rather as a debug info.

Layer logging from the watchdog was logging the integer value of a layer
instead of its string representation, so we change it.

SimonBrandner commented on pkg/conference/participant/tracker.go:
```diff
@@ -75,6 +76,16 @@ func (t *Tracker) RemoveParticipant(participantID ID) map[string]bool {
 	}
 
+	// Remove this participant's subscriptions.
+	for _, subscribers := range t.subscribers {
+		for subscriberID, subscription := range subscribers {
+			if subscriberID == participantID {
+				subscription.Unsubscribe()
+				delete(subscribers, subscriberID)
+			}
+		}
+	}
```
```suggestion
	for _, subscribers := range t.subscribers {
		for subscriberID, subscriber := range subscribers {
			if subscriberID == participantID {
				subscriber.Unsubscribe()
				delete(subscribers, subscriberID)
			}
		}
	}
```

daniel-abramov replied:
The thing is that the subscriber is **a key** in the map. subscription is **a value**. The actual type that is inside the value is a `Subscription`, that's why I also called the variable `subsscription`.

Perhaps the confusion comes from the fact that `subscribers` is **a map of maps**. `subscribers` is a map where the key is a track ID, while the value is a map that holds the correspondence between the subscriber ID (participant ID) and the subscription, i.e. contains both subscribers and subscriptions.

Or, if it would be simple to understand, `subscription` is semantically a map `map[TrackID]SubscriptionDetails`, where `SubscriptionDetails` is:

```go
type SubscriptionDetails struct {
    subscriber participant.ID
    subscription subscription.Subscription
}
```

Does it make any sense? 🙂

daniel-abramov replied:
In other words: a subscriber is a participant (or rather, its identifier) who is subscribed to something. Subscription is an actual entity (the actual subscription that handles the logic for the subscribed track, processes incoming packets and writes them to the output track).

SimonBrandner replied:
Ah, sorry I am confused

SimonBrandner commented on pkg/peer/subscription/audio.go:
```diff
@@ -0,0 +1,60 @@
+}
+
+func (s *AudioSubscription) readRTCP() {
+	for {
+		if _, _, err := s.sender.ReadRTCP(); err != nil {
+			if errors.Is(err, io.ErrClosedPipe) || errors.Is(err, io.EOF) {
+				return
+			}
+		}
+	}
```
This seems like wasting performence?

daniel-abramov replied:
Hm? What do you mean?

SimonBrandner replied:
The for loop doesn't seem to be doing anything - why is there?

daniel-abramov replied:
To read RTCPs. I don't know if it's required for audio, but it seems like it's generally something that should be there since reading RTCPs (even if we don't process them) makes Pion process the packet with the interceptors, which is required for NACKs, etc.

Or did you mean something different?

SimonBrandner replied:
Ah, so `ReadRTCP()` does something even though we don't use the packets it returns?

daniel-abramov replied:
> Ah, so ReadRTCP() does something even though we don't use the packets it returns?

Yeah, right. Pion has a concept of interceptors and uses "default interceptors" if none others are specified. This construction is very [common in Pion examples](https://github.com/pion/webrtc/blob/master/examples/reflect/main.go#L77).

> (are there any RTCP packets for audio tracks even?)

I'm not an expert to be honest, but I assume that things like NACKs (negative acknowledgments) for audio do still exist (i.e. so that the sender indicates packet loss), since we still have sequence numbers (and timestamps) for audio packets (both video and audio are transmitted as RTP streams).

SimonBrandner replied:
Argh, right, this feels weird tbh, Pion. Can we add a comment?

daniel-abramov replied:
Added!

SimonBrandner commented on pkg/peer/subscription/rtp.go:
```diff
@@ -0,0 +1,135 @@
+	// the **latest** (in terms of timestamp and number) packet. This is not
+	// necessarily the IDs of the **last forwarded packet**, due to packets
+	// that may arrive out-of-order.
+	maxOutgoingIDs PacketIdentifiers
```
While it might not be exactly the last forwarded packet, I still think `lastOutgoingIDs` would be more descriptive and give you a better picture when reading the code for the first time

SimonBrandner replied:
Personally, I prefer `last` even if slightly wrong... Does anyone else have any opinions on this?

daniel-abramov replied:
The thing is... I might need to use `last` as the last packet that we received when properly handling wrapping, so this name would be occupied 🙂 What about `latest`? Maybe sounds more reasonable than `last` and `max`? (I think I might actually indeed prefer `latest`)

SimonBrandner replied:
`latest` sounds great!

SimonBrandner commented on pkg/peer/subscription/rtp.go:
```diff
@@ -0,0 +1,135 @@
+		// If the incoming timeestamp or sequence number are smaller than the timestamp of the first packet after the switch,
+		// then we're getting the old packet (before switch), which we're not interested in, so we drop it.
+		if incomingIDs.timestamp < p.firstIncomingIDs.timestamp ||
+			incomingIDs.sequenceNumber < p.firstIncomingIDs.sequenceNumber {
```
Won't this break since they wrap around?

daniel-abramov replied:
Updated.

SimonBrandner commented on pkg/peer/remote_track.go:
```diff
@@ -0,0 +1,92 @@
+		}()
+
+		for {
+			index, _, readErr := remoteTrack.Read(rtpBuf)
```
Why not use `ReadRTP()`? It always felt a bit simpler to read to me

daniel-abramov replied:
I agree that `ReadRTP()` is a better choice when we're interested in RTP packets, that's also something that I use in when handling **video** tracks since we are working on a packet level when forwarding and rewriting packet IDs for the video.

In general, I think the only advantage of the `Read()` and `Write() over the `ReadRTP()` is that they use the buffer that you pass to the function, which means that `Read()` does not allocate, since we're reusing the same slice all over again (slight performance advantage). Since we don't really process audio packets in any way and just forward it within a single function, I just decided to use `Read()` / `Write()` here.

However, if you think that it harms the readability of this code, I can change it to the `ReadRTP()` and `WriteRTP()` (I think the performance loss, in this case, might be negligible anyway).

SimonBrandner commented on pkg/peer/messages.go:
```diff
@@ -18,16 +18,25 @@ type LeftTheCall struct {
+	// Information about the track (ID etc).
 	common.TrackInfo
+	// Simulcast configuration (can be `None` for non-simulcast tracks and for audio tracks).
+	Simulcast common.SimulcastLayer
```
I think either `Layer` or `SimulcastLayer` would be more clear on first read

SimonBrandner commented on pkg/peer/messages.go:
```diff
@@ -18,16 +18,25 @@ type LeftTheCall struct {
 
 type PublishedTrackFailed struct {
 	common.TrackInfo
+	Simulcast common.SimulcastLayer
```
I think either `Layer` or `SimulcastLayer` would be more clear on first read

SimonBrandner commented on pkg/peer/messages.go:
```diff
@@ -18,16 +18,25 @@ type LeftTheCall struct {
 type RTPPacketReceived struct {
 	common.TrackInfo
-	Packet *rtp.Packet
+	Simulcast common.SimulcastLayer
```
I think either `Layer` or `SimulcastLayer` would be more clear on first read

SimonBrandner commented on pkg/peer/messages.go:
```diff
@@ -48,4 +57,5 @@ type DataChannelAvailable struct{}
 
 type KeyFrameRequestReceived struct {
 	common.TrackInfo
+	Simulcast common.SimulcastLayer
```
I think either `Layer` or `SimulcastLayer` would be more clear on first read

SimonBrandner reviewed (commented):
Otherwise LGTM

daniel-abramov committed:
simulcast: expand rtp seqNums and timestamps

daniel-abramov committed:
subscription(audio): add a comment about RTCP read

daniel-abramov committed:
messages: rename `Simulcast` -> `SimulcastLayer`

daniel-abramov committed:
general: adapt code for the test job

After rebasing there are some things that the new PR must support in
order to compile well.

SimonBrandner reviewed (approved):
LGTM!
